### ISSUE TITLE
Dynamic URLs

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#3538]
'Dynamic URLs' is the best title I could think of to describe my idea.

Consider this scenario: Your site's URL structure is hackable, meaning each part of the URL between the slashes is an individual page, and they're structured hierarchly (e.g.`/about/staff/joe-bloggs` - `/about` is an overview page of the About section, `/about/staff/` is a views listing of Staff nodes, and `/about/staff/joe-bloggs` is the Staff node for Joe Bloggs). The problem is when you want to change 'About' to 'Who We Are' (e.g.`/who-we-are/staff/joe-bloggs`), or move the Staff page to another section of the site (e.g. `/contact-us/staff`). You then have to change the URL for the Staff views page, and the URL alias/pattern for each Staff node.

But what if the URL alias for Staff nodes was independent of where it was in the URL/site hierarchy? E.g. Staff nodes had a URL setting/pattern being `[firstname]-[surname]`, and then that was appended onto the URL of whatever the parent page was? And the URL for the Staff views page was simply `staff`, and that was appended onto the URL for whatever its parent page was.

There could maybe be a new admin page listing all URL paths that you could drag-and-drop (like the menu admin pages), and in doing so you'd be defining the site's URL structure.
E.g. You drag the Staff views page with URL `staff` under the About page with URL `about`, so the Staff page becomes accessible at `/about/staff`. Then you drag the URL pattern for Staff nodes (`[firstname]-[surname]`) under Staff and they become `/about/staff/[firstname]-[surname]`.

This way, if you ever wanted to change the structure around, you don't have to go renaming everything. You just drag the Staff page (and the Staff nodes under it) from About to Contact, and the URLs change automatically to `/contact/staff/[firstname]-[surname]`.

So basically we make it so that for any place where you enter a URL (for a node, taxonomy term, view, etc.), you're just defining the last part of the URL (after the final slash); the rest is generated dynamically based on where in the hierarchy that page/view/etc. sits.

I'm not sure if there are modules for Drupal that do this (though I doubt it based on how much it'd change how core works), and I'm not fully sure how it'd all work exactly. Just an idea I had that I'm throwing out there for discussion.

**Comment by klonos** [comment]
## Existing functionality

I think that (most of) this problem could be solved if instead of hard-coding certain parts of the path patterns, you'd use more "smart" path pattern **strategy**....

So instead of hard-coding:

`/about/staff/[node:field_firstname]-[node:field_surname]`

...you should do:

`about/[node:content-type:name]/[node:field_firstname]-[node:field_surname]`

Notice the use of `[node:content-type:name]` instead of just `[node:content-type]` which defaults to the machine name AFAIK. So if you renamed the content type from "Staff" to "Personnel", you'd still get a working pattern.

## New functionality that we could introduce

The only remaining bit to make "dynamic" then would be the leading `/about`. This could also be sorted if you used a "smarter" pattern...

So perhaps we could introduce a `[content:nid:path]` token. Say the "About" page is actually `/node/abc123`. You'd change the path pattern for "Staff"/"Personnel" nodes to:

`[content:nid:path]/[node:content-type:name]/[node:field_firstname]-[node:field_surname]`

For the views listing path, we'd need to use something like:

`[content:nid:path]/[content_type_machine_name:name]`

This of course would require:

- allowing tokens to be used in views displays paths
- introducing `[content_type_machine_name:*]` tokens

Of course, you could go the other way around and provide views-specific tokens. That way, you could use this pattern for the "Staff"/"Personnel" nodes:

`[content:nid:path]/[view_machine_name:display:path]/[node:field_firstname]-[node:field_surname]`

...

**Comment by klonos** [comment]
...anyway, the only problem after that would be to re-alias the existing nodes, which you can already do in Backdrop core with the bulk actions for aliases (`/admin/config/urls/path/bulk-update`). What we could do here in order to "connect the dots" is to detect alias patterns changes, and then throw some sort of `info` type of message along the lines of:

`The path pattern for xyz was updated. New content will be created using the new pattern. Old content has not been updated, so will retain its old path. To update the old content, you can use the [URL aliases bulk actions](url).`

**Comment by klonos** [comment]
> ...Notice the use of `[node:content-type:name]` instead of just `[node:content-type]`

I have not actually tested this ^^, so we might need something like https://www.drupal.org/project/token_modifier in order to support lower-casing the content type name:

`[token-modifier:lowercase:node:content-type:name]`

**Comment by klonos** [comment]
...looks like we need https://www.drupal.org/project/aet either ported in contrib for things like that, or merged into core so that these kinds of tokens are natively supported. It supports things like `[aet:node:<nid>]` or `[aet:node:<nid>:body]`.

**Comment by olafgrabienski** [comment]
In some cases, the token `[node:menu-link:parents]` can also be of interest.

**Comment by BWPanda** [comment]
Sorry I didn't get around to replying to this. But yes, it seems that using tokens in a planned and structured way pretty much achieves what I was wanting here. Thanks guys!
